File: Spring/les/src/main/java/com/jga/les/model/CompraProduto.java
package com.jga.les.model;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonManagedReference;

import jakarta.persistence.EmbeddedId;
import jakarta.persistence.Entity;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.MapsId;
import lombok.Data;

@Data
@Entity
public class CompraProduto {
    @EmbeddedId
    private CompraProdutoKey id;
    
    @ManyToOne
    @MapsId("idcompra")
    @JoinColumn(name = "compra_id")
    @JsonBackReference
    private Compra compra;

    @ManyToOne
    @MapsId("idproduto")
    @JoinColumn(name = "produto_id")
    @JsonManagedReference  
    private Produto produto;

    private Double qntd;
    private Double preco;
    private Double custo;
}

File: Flutter/les/lib/model/compra/compra_produto.dart
import 'package:les/model/entity.dart';
import 'package:les/model/produto/produto.dart';

class Id implements Entity{
  int idCompra;
  int idProduto;

  Id({
    required this.idCompra,
    required this.idProduto
  });

  @override
  Map<String, dynamic> toJson() {
    return {
      'idcompra': idCompra,
      'idproduto': idProduto
    };
  }

  factory Id.fromJson(Map<String, dynamic> json){
    return Id(
      idCompra: json['idcompra'],
      idProduto: json['idproduto']
    );

  }

  @override
  int? id;
}

class CompraProduto implements Entity{
  Id idComposto;
  Produto produto;
  double qntd;
  double preco;
  double custo;

  CompraProduto({
    required this.idComposto,
    required this.produto,
    required this.qntd,
    required this.preco,
    required this.custo
  });

  factory CompraProduto.fromJson(Map<String, dynamic> json){
    return CompraProduto(
      idComposto: Id.fromJson(json['id']),
      produto: Produto.fromJson(json['produto']),
      qntd: json['qntd'],
      preco: json['preco'],
      custo: json['custo']
    );
  }

  @override
  Map<String, dynamic> toJson(){
    return {
      'id': idComposto.toJson(),
      'produto': produto.toJson(),
      'qntd': qntd,
      'preco': preco,
      'custo': custo
    };
  }

  @override
  int? id;

}